esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -1881,49 +1881,159 @@ IPropertyTree *CFileSprayEx::getAndValidateDropZone(const char *path, const char
+}
+
+static void ensureDZFileScopePermission(IEspContext& context, const char* plane, const char* host, const char* path, SecAccessFlags permissionReq)
+{
+    SecAccessFlags permission = getDZFileScopePermissions(context, plane, path, host);
+    if (permission < permissionReq)
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Access DropZone Scope %s %s not allowed for user %s (permission:%s). Read Access Required.",
+            plane, path, context.queryUserId(), getSecAccessFlagName(permission));
+}
+
+void CFileSprayEx::readAndCheckSpraySourceReq(IEspContext& context, const char* srcHost, const char* srcPath, const char* srcPlane,
+    StringBuffer& sourcePlaneReq, StringBuffer& sourceHostReq, StringBuffer& sourcePathReq)
+{
+    //Clean srcPlane and srcHost
+    sourcePlaneReq.set(srcPlane);
+    sourceHostReq.set(srcHost);
+    sourcePlaneReq.trim();
+    sourceHostReq.trim();
+    if (sourcePlaneReq.isEmpty() && sourceHostReq.isEmpty())
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Neither source plane nor source host supplied. A dropzone name specified in the SourcePlane is preferred.");
+
+    //Clean srcPath
+    StringBuffer sourcePath, s(srcPath);
+    s.trim();
+    if (s.isEmpty())
+        throw makeStringException(ECLWATCH_INVALID_INPUT, "Source path for spraying a file not specified.");
+    if (!sourcePlaneReq.isEmpty())
+        getStandardPosixPath(sourcePath, s.str());
+    else
+        getStandardPosixPath(sourcePathReq, s.str());
+
+    StringArray fileReqs;
+    fileReqs.appendList(sourcePath, ","); // handles comma separated files
+
+    if (sourcePlaneReq.isEmpty())
+    {
+        checkSpraySourcePathAndGetPlane(context, sourceHostReq, fileReqs, sourcePlaneReq);
+        return;
+    }
+
+    //Validate sourcePlaneReq
+    Owned<IPropertyTree> dropZone = getDropZonePlane(sourcePlaneReq);
+    if (!dropZone)
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown dropzone: %s", sourcePlaneReq.str());
+
+    //Validate sourceHostReq
+    SocketEndpoint sourceHostEp;
+    if (!sourceHostReq.isEmpty())
+    {
+        if (!isHostInPlane(dropZone, sourceHostReq, true))
+            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Host %s is not defined within the dropzone %s.", sourceHostReq.str(), sourcePlaneReq.str());
+        sourceHostEp.set(sourceHostReq);
+    }
+
+    //Validate fileReqs
+    const char* dropZonePlanePath = dropZone->queryProp("@prefix");
+    ForEachItemIn(i, fileReqs)
+    {
+        const char* fileReq = fileReqs.item(i);
+        if (isEmptyString(fileReq))
+            continue;
+
+        //Parse the fileReq to find/validate possible host/ip and local path.
+        StringBuffer hostInPath, localPath, absPath;
+        checkHostInSpraySourcePath(fileReq, sourcePlaneReq, dropZone, sourceHostReq, &sourceHostEp, localPath, hostInPath);
+
+        //Validate file path.
+        const char* path = localPath.isEmpty() ? fileReq : localPath.str();
```

Comment:
would be clearer if checkHostInSpraySourcePath had returned true if was a UNC path and had validated, and do, e.g.:
```
const char *path;
if (IsHostInSprayPath(fileReq, sourcePlaneReq, dropZone, sourceHostReq, &sourceHostEp, localPath, hostInPath))
    path = localPath.str();
else
   path = fileReq;